test/utils/structs/Vector.test.js

```diff
@@ -0,0 +1,91 @@
+const { expectEvent, expectRevert } = require('@openzeppelin/test-helpers');
+const { expect } = require('chai');
+
+const Bytes32VectorMock = artifacts.require('Bytes32VectorMock');
+
+async function getContent (vector) {
+  const length = await vector.length().then(bn => bn.toNumber());
+  const values = await Promise.all(Array(length).fill().map((_, i) => vector.at(i)));
+  return values;
+}
+
+contract('Vector', function (accounts) {
+  // Bytes32Vector
+  describe('EnumerableVector', function () {
+    const bytesA = '0xdeadbeef'.padEnd(66, '0');
+    const bytesB = '0x0123456789'.padEnd(66, '0');
+    const bytesC = '0x42424242'.padEnd(66, '0');
+    const bytesD = '0x171717'.padEnd(66, '0');
+
+    beforeEach(async function () {
+      this.vector = await Bytes32VectorMock.new();
+    });
+
+    it('empty vector', async function () {
+      expect(await this.vector.empty()).to.be.equal(true);
+      expect(await getContent(this.vector)).to.have.all.members([]);
+    });
+
+    describe('with content', function () {
+      beforeEach(async function () {
+        await this.vector.pushBack(bytesB);
+        await this.vector.pushFront(bytesA);
+        await this.vector.pushBack(bytesC);
+      });
+
+      it('content is valid', async function () {
+        expect(await this.vector.empty()).to.be.equal(false);
+        expect(await this.vector.front()).to.be.equal(bytesA);
+        expect(await this.vector.back()).to.be.equal(bytesC);
+        expect(await getContent(this.vector)).to.have.all.members([ bytesA, bytesB, bytesC ]);
+      });
+
+      it('out of bound access', async function () {
+        await expectRevert.unspecified(this.vector.at(10));
+      });
+
+      describe('push', function () {
+        it('front', async function () {
+          await this.vector.pushFront(bytesD);
+
+          expect(await this.vector.front()).to.be.equal(bytesD);
+          expect(await this.vector.back()).to.be.equal(bytesC);
+          expect(await getContent(this.vector)).to.have.all.members([ bytesD, bytesA, bytesB, bytesC ]);
```

Comment:
Would be nice to write the tests in terms of JavaScript array operations. So we could check that the array is now `initialArray.unshift(bytesD)`, and so on.